### PR TITLE
[TT-8518]Added enterprise portal to tyk-charts

### PR DESCRIPTION
**Pull request by andrei-tyk**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)
- [ ] Documentation updates or improvements.

## Checklist

- [ ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If PRing from your fork, don't come from your `master`!
- [ ] Make sure you are making a pull request against our **`master` branch** (left side). Also, it would be best if you started *your change* off *our latest `master`*.
- [ ] My change requires a change to the documentation.
     - [ ] I have manually updated the README(s)/documentation accordingly.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: components/tyk-enterprise-portal/Chart.yaml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+# This is the chart version. This version number should be incremented each time you make changes
+# to the chart and its templates, including the app version.
+# Versions are expected to follow Semantic Versioning (https://semver.org/)
+version: 1.0.0-beta4
+
+# This is the version number of the application being deployed. This version number should be
+# incremented each time you make changes to the application. Versions are not expected to
+# follow Semantic Versioning. They should reflect the version the application is using.
+# It is recommended to use it with quotes.
+appVersion: "5.0.0"
```
Comment (komalsukhani) [inline components/tyk-enterprise-portal/Chart.yaml:24]: Does this represent portal version?

File: components/tyk-bootstrap/templates/bootstrap-post-install.yaml [hunk 2]
```diff
@@ -80,11 +81,11 @@ spec:
             - name: OPERATOR_SECRET_NAME
               value: "tyk-operator-conf"
             {{- end }}
-            {{- if .Values.bootstrap.enterprisePortalSecret }}
+            {{- if .Values.global.components.enterprisePortal }}
```
Comment (komalsukhani) [inline components/tyk-bootstrap/templates/bootstrap-post-install.yaml:84]: Now portal secret will always be created right?
Previously, `bootstrap.enterprisePortalSecret` determined is secret needs to be created or not.

File: components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+kind: StatefulSet
+metadata:
+  name: {{ include "tyk-enterprise-portal.fullname" . }}
+  labels:
+    app: {{ include "tyk-enterprise-portal.fullname" . }}
+    chart: {{ include "tyk-enterprise-portal.chart" . }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+spec:
+  serviceName: service-{{ include "tyk-enterprise-portal.fullname" . }}
```
Comment (komalsukhani) [inline components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml:11]: Name doesn't match with the service name defined in `service-enterprise-portal.yaml`

File: components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+      app: {{ include "tyk-enterprise-portal.fullname" . }}
+      release: {{ .Release.Name }}
+  template:
+    metadata:
+      labels:
+        app: {{ include "tyk-enterprise-portal.fullname" . }}
+        release: {{ .Release.Name }}
+      annotations:
+        checksum/secrets: "{{ include (print $.Template.BasePath "/secret.yaml") . | sha256sum }}"
+        {{- if .Values.mounts }}
```
Comment (komalsukhani) [inline components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml:27]: This is not required anymore as we don't want to support mounts.

File: components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+                {{ end }}
+                key: EnterprisePortalStorgeConnectionString
+          - name: PORTAL_THEMING_THEME
+            value: "default"
+          - name: PORTAL_THEMING_PATH
+            value: "./themes"
+          - name: PROVIDER_NAME
+            value: "tyk"
+          - name: TYK_URL
+            value: "http{{ if .Values.global.tls.enterprisePortal }}s{{ end }}://dashboard-svc-{{ .Release.Name }}.{{ .Release.Namespace }}.svc:{{ .Values.global.servicePorts.dashboard }}"
```
Comment (komalsukhani) [inline components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml:99]: Shouldn't it be 
```suggestion
            value: "http{{ if .Values.global.tls.dashboard }}s{{ end }}://dashboard-svc-{{ .Release.Name }}.{{ .Release.Namespace }}.svc:{{ .Values.global.servicePorts.dashboard }}"
```

File: components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+          - name: TYK_AUTH
+            valueFrom:
+              secretKeyRef:
+                {{ if .Values.secrets.useSecretName }}
+                name: {{ .Values.secrets.useSecretName }}
+                {{ else if .Values.global.bootstrap.enterprisePortal}}
+                name: {{ .Values.global.secrets.enterprisePortal}}
+                {{ else }}
+                name: secrets-{{ include "tyk-enterprise-portal.fullname" . }}
```
Comment (komalsukhani) [inline components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml:108]: `TYK_AUTH` key is not set in the secret of enterprise portal.

File: components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+                key: TYK_AUTH
+          - name: TYK_ORG
+            valueFrom:
+              secretKeyRef:
+                {{ if .Values.secrets.useSecretName }}
+                name: {{ .Values.secrets.useSecretName }}
+                {{ else if .Values.global.bootstrap.enterprisePortal}}
+                name: {{ .Values.global.secrets.enterprisePortal}}
+                {{ else }}
+                name: secrets-{{ include "tyk-enterprise-portal.fullname" . }}
```
Comment (komalsukhani) [inline components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml:119]: `TYK_ORG` is not set in the secret of enterprise portal.

File: components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+            subPath: tyk-portal
+          - name: enterprise-portal-pvc-{{ include "tyk-enterprise-portal.fullname" . }}
+            mountPath: /opt/portal/themes
+            subPath: themes
+          {{- if eq .Values.storage.database.type "sqlite3" }}
+          - name: enterprise-portal-pvc-{{ include "tyk-enterprise-portal.fullname" . }}
+            mountPath: /opt/portal/db
+            subPath: db
+          {{- end }}
+          {{- if .Values.mounts }}
```
Comment (komalsukhani) [inline components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml:170]: Not needed anymore.

File: components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+          initialDelaySeconds: {{ default 60 .Values.readinessProbe.initialDelaySeconds}}
+          periodSeconds: 10
+          timeoutSeconds: 3
+          failureThreshold: 3
+      {{- if .Values.securityContext }}
+      securityContext:
+      {{- toYaml .Values.securityContext | nindent 8 }}
+      {{- end}}
+      volumes:
+        {{- if .Values.mounts }}
```
Comment (komalsukhani) [inline components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml:199]: Not needed anymore.

File: components/tyk-enterprise-portal/values.yaml [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+##       defaultMode: 420
+##       secretName: ca-certs
+extraVolumes: []
+## extraVolumeMounts A list of volume mounts to be added to the pod
+## extraVolumeMounts:
+##   - name: ca-certs
+##     mountPath: /etc/ssl/certs/ca-certs.crt
+##     readOnly: true
+extraVolumeMounts: []
+mounts: []
```
Comment (komalsukhani) [inline components/tyk-enterprise-portal/values.yaml:90]: This is not needed anymore

File: components/tyk-enterprise-portal/values.yaml [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+
+
+secrets:
+  # If you don't want to store plaintext secrets in the Helm value file and would
+  # rather provide the k8s Secret externally please populate the value below
+  useSecretName: ""
```
Comment (komalsukhani) [inline components/tyk-enterprise-portal/values.yaml:96]: Can you move it to global section and also specify in the comments what fields can be set in the secret?

File: components/tyk-enterprise-portal/values.yaml [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+
+livenessProbe:
+  initialDelaySeconds: 60
+
+readinessProbe:
+  initialDelaySeconds: 60
+
+
+global:
```
Comment (komalsukhani) [inline components/tyk-enterprise-portal/values.yaml:105]: Can you move it at the top like in other components? Just for consistency.

**Review by komalsukhani — CHANGES_REQUESTED** [review]

File: components/tyk-bootstrap/values.yaml [hunk 13]
```diff
@@ -5,6 +5,18 @@ global:
...
     # The license key needed for the Tyk Dashboard.
     dashboard: ""
+  adminUser:
+    firstName: admin
+    lastName: user
+    email: default@example.com
+    # Set a password or a random one will be assigned. Admin user password must have at least 6 characters.
+    password: "123456"
+  secrets:
+    enterprisePortal: tyk-enterprise-portal-conf
```
Comment (buraksekili) [inline components/tyk-bootstrap/values.yaml:15]: can you please write a comment on what this secret means?

File: components/tyk-enterprise-portal/values.yaml [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+# Enterprise Portal will only bootstrap if the tyk-enterprise-portal-conf secret
+# is present. The secret is created during the dashboard bootstrapping and
+# be turned on and off using the dash.enterprisePortalSecret bool.
+
+
+# The hostname to bind the Enterprise Portal to.
+hostName: tyk-enterprise-portal.local
+# Enterprise Portal license.
+license: ""
+# Enterprise Portal admin information.
```
Comment (buraksekili) [inline components/tyk-enterprise-portal/values.yaml:14]: i guess this comment does not belong here?

File: components/tyk-enterprise-portal/values.yaml [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+  initialDelaySeconds: 60
+
+readinessProbe:
+  initialDelaySeconds: 60
+
+
+global:
+  adminUser:
+    email: default@example.com
+    password: A8vMlXSWY3
```
Comment (buraksekili) [inline components/tyk-enterprise-portal/values.yaml:108]: why do we use this password as a default one? can't we use `123456` as we used in dashboard? it looks confusing

File: tyk-single-dc/values.yaml [hunk 16]
```diff
@@ -5,12 +5,22 @@
...
+    firstName: admin
+    lastName: user
+    email: default@example.com
+    # Set a password or a random one will be assigned.
+    password: "123456"
   components:
     # Determines whether or not the pump component should be installed.
     pump: true
-
+    enterprisePortal: true
```
Comment (buraksekili) [inline tyk-single-dc/values.yaml:18]: will the enterprise portal be enabled by default? just asking.

File: components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+          - name: PORTAL_LICENSEKEY
+            valueFrom:
+              secretKeyRef:
+                {{ if .Values.secrets.useSecretName }}
+                name: {{ .Values.secrets.useSecretName }}
+                {{ else }}
+                name: secrets-{{ include "tyk-enterprise-portal.fullname" . }}
+                {{ end }}
+
+                key: EnterprisePortalLicense
```
Comment (buraksekili) [inline components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml:67]: Can't we make this field configurable?

File: components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml (hunk 5 above)
Comment (buraksekili) [reply components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml:99]: Yes, agree. Also, this TYK_URL value seems missing to me. When we install `tyk-single-dc` chart, the dashboard svc has the following name:
```
dashboard-svc-tyk-single-dc-tyk-dashboard.tyk.svc
```
However, this TYK_URL has no `tyk-dashboard` suffix, which is the name of the dashboard subchart.

File: components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml (hunk 3 above)
Comment (buraksekili) [reply components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml:11]: Agree. Due to this, service has no control over pods, so we cannot access portal through service name (or port-forward)

File: components/tyk-enterprise-portal/templates/service-enterprise-portal.yaml [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+{{- end }}
+spec:
+  ports:
+  - port: {{ .Values.service.port }}
+    targetPort: {{ .Values.containerPort }}
+    protocol: TCP
+    name: http
+  type: {{ .Values.service.type }}
+  selector:
+    app: enterprise-portal-{{ include "tyk-enterprise-portal.fullname" . }}
```
Comment (buraksekili) [inline components/tyk-enterprise-portal/templates/service-enterprise-portal.yaml:24]: This label is incorrect. It should be
```suggestion
    app: {{ include "tyk-enterprise-portal.fullname" . }}
```

File: components/tyk-enterprise-portal/Chart.yaml (hunk 1 above)
Comment (andrei-tyk) [reply components/tyk-enterprise-portal/Chart.yaml:24]: No, tyk-version, will update

File: components/tyk-bootstrap/templates/bootstrap-post-install.yaml (hunk 2 above)
Comment (andrei-tyk) [reply components/tyk-bootstrap/templates/bootstrap-post-install.yaml:84]: Changed to  .Values.global.bootstrap.enterprisePortal

File: components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml (hunk 3 above)
Comment (andrei-tyk) [reply components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml:11]: Fixed

File: components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml (hunk 5 above)
Comment (andrei-tyk) [reply components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml:99]: Fixed

File: components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml (hunk 7 above)
Comment (andrei-tyk) [reply components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml:119]: Fixed

File: components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml (hunk 8 above)
Comment (andrei-tyk) [reply components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml:170]: Fixed

File: components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml (hunk 9 above)
Comment (andrei-tyk) [reply components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml:199]: Fixed

File: components/tyk-enterprise-portal/values.yaml (hunk 10 above)
Comment (andrei-tyk) [reply components/tyk-enterprise-portal/values.yaml:90]: Fixed

File: components/tyk-enterprise-portal/values.yaml (hunk 11 above)
Comment (andrei-tyk) [reply components/tyk-enterprise-portal/values.yaml:96]: Fixed

File: components/tyk-enterprise-portal/values.yaml (hunk 12 above)
Comment (andrei-tyk) [reply components/tyk-enterprise-portal/values.yaml:105]: Fixed

File: components/tyk-enterprise-portal/values.yaml (hunk 14 above)
Comment (andrei-tyk) [reply components/tyk-enterprise-portal/values.yaml:14]: Fixed

File: components/tyk-enterprise-portal/values.yaml (hunk 15 above)
Comment (andrei-tyk) [reply components/tyk-enterprise-portal/values.yaml:108]: Fixed

File: tyk-single-dc/values.yaml (hunk 16 above)
Comment (andrei-tyk) [reply tyk-single-dc/values.yaml:18]: I set it to false and we'll discuss in next meeting.

File: components/tyk-enterprise-portal/templates/service-enterprise-portal.yaml (hunk 18 above)
Comment (andrei-tyk) [reply components/tyk-enterprise-portal/templates/service-enterprise-portal.yaml:24]: Fixed

File: components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml (hunk 17 above)
Comment (andrei-tyk) [reply components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml:67]: Why should we? It's something documented and would just add unnecessary clutter to the yaml

File: components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml [hunk 19]
```diff
@@ -0,0 +1,211 @@
...
+                {{ end }}
+                key: EnterprisePortalStorgeConnectionString
+          - name: PORTAL_THEMING_THEME
+            value: "default"
+          - name: PORTAL_THEMING_PATH
+            value: "./themes"
+          - name: PROVIDER_NAME
+            value: "tyk"
+          - name: TYK_URL
+            value: "http{{ if .Values.global.tls.dashboard }}s{{ end }}://dashboard-svc-{{ .Release.Name }}-tyk-dashboard.{{ .Release.Namespace }}.svc:{{ .Values.global.servicePorts.dashboard }}"
```
Comment (buraksekili) [inline components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml:93]: what happens if the dev port is installed as a subchart? I mean without installing it via an umbrella chart? Will the `tyk-dashboard` suffix appear?

File: components/tyk-bootstrap/templates/bootstrap-post-install.yaml [hunk 20]
```diff
@@ -29,18 +29,19 @@ spec:
       serviceAccountName: k8s-bootstrap-role
       containers:
         - name: bootstrap-tyk-post-install
-          image: tykio/tyk-k8s-bootstrap-post:latest
+          {{/*  emergency solution so it will work with the new k8s bootstrap modifications*/}}
+          image: andreiattyk/tyk-repo:post33
```
Comment (buraksekili) [inline components/tyk-bootstrap/templates/bootstrap-post-install.yaml:33]: can't we publish officially yet?

File: components/tyk-bootstrap/values.yaml [hunk 21]
```diff
@@ -27,6 +38,7 @@ global:
     dashboard: false
     # When true, sets the gateway protocol to HTTPS.
     gateway: false
+    key: EnterprisePortalLicense
```
Comment (buraksekili) [inline components/tyk-bootstrap/values.yaml:41]: what does this field stand for? Can you add a comment?

File: components/tyk-bootstrap/templates/bootstrap-post-install.yaml (hunk 20 above)
Comment (andrei-tyk) [reply components/tyk-bootstrap/templates/bootstrap-post-install.yaml:33]: We can discuss releasing the k8s-bootstrap update today

File: components/tyk-bootstrap/values.yaml (hunk 21 above)
Comment (andrei-tyk) [reply components/tyk-bootstrap/values.yaml:41]: This was an useless piece of code. Deleted.

File: components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml (hunk 19 above)
Comment (buraksekili) [reply components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml:91]: makes sense. let's make it configurable as you suggested. As a default value, I think we can use `"http{{ if .Values.global.tls.dashboard }}s{{ end }}://dashboard-svc-{{ .Release.Name }}-tyk-dashboard.{{ .Release.Namespace }}.svc:{{ .Values.global.servicePorts.dashboard }}"`

If users installs component developer portal, they can change it

Comment (andrei-tyk) [reply components/tyk-enterprise-portal/templates/statefulset-enterprise-portal.yaml:94]: Added required change

File: components/tyk-bootstrap/templates/bootstrap-post-install.yaml [hunk 22]
```diff
@@ -29,18 +29,19 @@ spec:
       serviceAccountName: k8s-bootstrap-role
       containers:
         - name: bootstrap-tyk-post-install
+          {{/*  emergency solution so it will work with the new k8s bootstrap modifications*/}}
```
Comment (buraksekili) [inline components/tyk-bootstrap/templates/bootstrap-post-install.yaml:32]: We can delete this comment :)

**Review by komalsukhani — APPROVED** [review]

**Pull request merged**